### DISCUSSION TITLE
Render props in a <script type="application/json"> rather than an escaped attribute?

### DISCUSSION THREAD

**Discussion by spacedawwwg** [withastro/roadmap#635] [Proposal]
### Body

# Summary

Render props in a `<script id="astro-island-props-1emU40" type="application/json">` rather than an escaped attribute. Positive when lots of props are passed reducing markup and making props more readable in source code

# Background & Motivation

I have been developing a mega-menu using a Vue component. The component requires the menu items passed as props generating a large `props` attribute

![image](https://github.com/withastro/roadmap/assets/1691867/3e5b975a-e236-44c5-8564-311325bd1cd6)

That's a LOT of slashes and `&quot;`. I feel this would be much neater in a script tag, removing any need to escape.

```HTML
<script id="astro-island-props-1emU40" type="application/json">
{"items":[{"href":"/","text":"Example"}]} // etc. etc.
</script>
```

```JS
const props = JSON.parse(document.getElementById(`astro-island-props-${uid}`).textContent);
```

As props are not cached (they are served in the HTML) it'd be nice to keep there output as small as possible, saving as many bytes as possible.

# Goals

A **concise, bulleted-list** outlining the intended goals of this RFC. 

- reducing markup
- making props more readable in source code

**Comment by natemoo-re** [comment]
Yeah, I'm very in favor of making this change! One problem is that we'll have to inline the `script` inside of the `astro-island` itself since we're streaming out the response. I guess `astro-island` already breaks sibling selectors so it's not a huge problem.

**Reply by spacedawwwg** [reply]
So, you're thinking something that looks a little like this?
```HTML
<astro-island uid="1iat0c" component-url="/assets/client.77d92ba7.js" component-export="default" renderer-url="/assets/Button.ac32abea.js" ssr="" client="visible">
  <script type="application/json" id="astro-island-props-1iat0">
    {"label":"The label"}
  </script>
  <button>The label</button>
</astro-island>
```

**Reply by natemoo-re** [reply]
Yep, that seems reasonable! Unfortunately @belluzj pointed out that JSON is probably a terrible format for this since the quotes become increasingly more escaped. Working on a solution in [this `astro` bug report](https://github.com/withastro/astro/issues/7978#issuecomment-1668248601).

**Comment by belluzj** [comment]
We're also using Astro with deeply nested data structures to pass to the client and seeing the same "bloat" from the serialization. I believe that in parallel with this proposition, even the data that gets put in the attribute could be much leaner, because it looks like perfectly normal nested JSON data gets repeatedly JSON.stringfy-ed, creating a seemingly quadratic amount of quote escaping, on top of the use of XML entities for the quotes. I will create an example repository to try and reproduce the issue.

**Reply by belluzj** [reply]
I posted an issue here: https://github.com/withastro/astro/issues/7978

I'm willing to look into contributing a fix, if someone from the project could hint at which file does the prop serialization and deserialization for islands.

**Reply by natemoo-re** [reply]
Posted an in-depth comment on https://github.com/withastro/astro/issues/7978#issuecomment-1668248601! Thanks a lot of opening an issue to point out this problem and being willing to submit a PR! Exploring a new serialization approach seems totally reasonable to mitigate this problem.
